src/plugins/data/public/actions/abort_data_query_action.ts

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ActionByType, createAction } from '../../../ui_actions/public';
+
+export const ACTION_ABORT_DATA_QUERY = 'ACTION_ABORT_DATA_QUERY';
+
+export interface AbortDataQueryContext {
+  reason: string;
+}
+
+// Create the action creator function
+export function createAbortDataQueryAction(
+  abortControllerRef: React.MutableRefObject<AbortController | undefined>
+): ActionByType<typeof ACTION_ABORT_DATA_QUERY> {
+  return createAction<typeof ACTION_ABORT_DATA_QUERY>({
+    type: ACTION_ABORT_DATA_QUERY,
+    id: ACTION_ABORT_DATA_QUERY,
+    shouldAutoExecute: async () => true,
+    execute: async (context: AbortDataQueryContext) => {
+      try {
+        if (abortControllerRef.current) {
+          // Abort existing query
+          abortControllerRef.current.abort();
+        }
```

Comment:
nit: we can move if condition outside of try like `if (!abortControllerRef.current) return` for better readability